### DISCUSSION TITLE
MVLegendImageClass usage

### DISCUSSION THREAD

**Discussion by mlandsfeld** [tethysplatform/tethys#660] [Q&A]
I am trying to use an image for the legend in my MVLayer and not having success. Here is my attempt:

        geoserver_layer = MVLayer(
            source='ImageWMS',
            options={
                'url': 'https://....../geoserver/wms',
                'params': {'LAYERS': selected_layer},
                'serverType': 'geoserver'
            },
            legend_title=legend_title,
            legend_extent=[-114, 36.5, -109, 42.5],
            legend_classes=[
              MVLegendImageClass(
            	value='Precipitation(mm)',
            	image_url='https://....../images/legends/precip_dekad_data_raster.png'
              )
			]
		)

Any help appreciated!

**Comment by swainn** [comment]
@mlandsfeld it appears to me that you are using it correctly. Please provide answers to the following questions to help get to the bottom of this:

1. Is the image URL publicly accessible? Enter it in your web browser to test it.
2. Does the URL return an image?
3. Are there any errors in the JavaScript console in your web browser when you access the page? Press `F12` and select the Console tab. Then refresh the page to see if any errors appear.
4. Are there any network errors while retrieving the image. With the web console open, select the Network tab and refresh the page. Look for the URL for the legend image and see if there are any errors.

**Reply by mlandsfeld** [reply]
Hi Nathan,
The image is publicly accessible as you can see here:

https://chc-ewx2.chc.ucsb.edu/images/legends/precip_dekad_data_raster.png

But there is a cross origin issue with using the GeoServer on that site and retrieving this image. This warning is given in the console/issues pane in Chrome.

Because a cookie's SameSite attribute was not set or is invalid, it defaults to SameSite=Lax, which prevents the cookie from being sent in a cross-site request. This behavior protects the user data from accidentally leaking to third parties and cross-site request forgery. 

Resolve this issue by updating the attributes of the cookie:
-- specify SameSite=None and Secure if the cookies should be sent in cross-site requests. This enables third-party use.
-- Specify SameSite=Strict if the cookie should not be sent in the cross-site requests._

These are images being used by one of our data viewers and are being served with nginx. I could pull these in on my local machine and use them that way but I would like to use those that are on the viewer's server. Any suggestions? Thanks!

**Comment by swainn** [comment]
@mlandsfeld thank you for the update. I'm not familiar with this particular issue, so I'll need to take time to see if I can reproduce the issue and investigate more to see if there is an adjustment that needs to be made to Tethys. Part of the confusion with CORS issues recently is that web browsers are changing their default behavior to be more strict to protect users.

 Is this error occurring while you are developing a Tethys app (i.e.: you are using `tethys manage start` and accessing testing locally with `localhost:8000`/`127.0.0.1:8000`?

**Comment by mlandsfeld** [comment]
I am running the Tethys app locally and using tethys manage start but accessing a remote geoserver at:
https://chc-ewx2.chc.ucsb.edu:8443/geoserver/web
for my data layers. I am retrieving the legend images on that same server. I checked that Tomcat was allowing CORS, and it is set up for that, but I am getting a Referrer Policy: strict-origin-when-cross-origin returned. I have the CorsFilter set up like this:

```
<filter>
    <filter-name>CorsFilter</filter-name>
    <filter-class>org.apache.catalina.filters.CorsFilter</filter-class>
    <init-param>
        <param-name>cors.allowed.origins</param-name>
        <param-value>*</param-value>
    </init-param>
    <init-param>
        <param-name>cors.allowed.headers</param-name>
        <param-value>Content-Type,X-Requested-With,accept,Origin,Access-Control-RequestMethod,Access-Control-Request-Headers,Authorization,dsName</param-value>
    </init-param>
</filter>

<filter-mapping>
    <filter-name>CorsFilter</filter-name>
    <url-pattern>/*</url-pattern>
</filter-mapping>
```

I looked to for any references to changes to this method but have not found any. Thanks!

**Comment by mlandsfeld** [comment]
Sorry, that CorsFilter XML snippet didn't translate well in my last post. It was a copy and paste from my web.xml file. If you would want to see that I can send a screenshot of it.

But something I just noticed is that I have a {% csrf_token %} Django tag in my map.html template for my map page. I started from the geoserver_app Tethys template and it came with that included. If I remove the tag I get a 403 Forbidden error. Is there an alternate to that tag?

**Reply by swainn** [reply]
The `{% csrf_token %}` is needed when you want to submit a POST request to the server (say in a form). The convention is to use the POST method on requests that are potentially "destructive" (delete, create, or change data on your server). The csrf_token basically verifies to the server that the request is coming from a page that it generated and prevents outsiders from submitting malicious requests. 


TL;DR: The server will not accept a POST request without a `csrf_token` (hence the 403 Forbidden error).
